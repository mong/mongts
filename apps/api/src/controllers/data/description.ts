import { RequestHandler, Request } from "express";
import { descriptionModel, Filter } from "../../models/data";

export const descriptionController: RequestHandler = async (req, res) => {
  try {
    const register: string = req.params.register;
    const query = parseQuery(req);
    const rows = await descriptionModel(register, query.filter);
    res.json(rows);
  } catch (error) {
    const error_message =
      error instanceof Error ? error.message : String(error);
    res.status(500).json({ message: error_message });
  }
};

interface Query {
  filter?: Filter;
}

function parseQuery(req: Request): Query {
  const query: Query = {};

  if (typeof req.query === "object" && !Array.isArray(req.query)) {
    query.filter = {};

    if (typeof req.query.type === "string") {
      query.filter.type = req.query.type;
    }
  }
  return query;
}
