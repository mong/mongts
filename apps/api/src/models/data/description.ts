import { Knex } from "knex";
import db from "../../db";
import { Filter } from ".";

export interface Description {
  id: string;
  dg_id: string | null;
  include: number | null;
  title: string | null;
  name: string | null;
  type: string | null;
  min_denominator: number | null;
  level_green: number | null;
  level_yellow: number | null;
  level_direction: number | null;
  short_description: string | null;
  long_description: string | null;
  registry_id: number;
  rname: string | null;
  full_name: string;
}

export const descriptionModel = (
  register: string,
  filter: Filter | undefined
): Promise<Description[]> =>
  db
    .select("ind.*", "registry.name as rname", "registry.full_name")
    .from("ind")
    .leftJoin("registry", "ind.registry_id", "registry.id")
    .modify(withFilter, filter)
    .where("registry_id", function (this: Knex) {
      this.select("id").from("registry").where("name", register);
    });

function withFilter(builder: Knex.QueryBuilder, filter?: Filter) {
  if (filter?.type) {
    if (filter.type === "dg") {
      builder.where("type", filter.type).orWhereNull("type");
    } else if (filter.type === "ind") {
      builder.whereNot("type", "dg").whereNotNull("type");
    }
  }
}
