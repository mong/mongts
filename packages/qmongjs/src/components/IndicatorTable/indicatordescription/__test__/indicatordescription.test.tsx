import { render } from "@testing-library/react";

import { IndicatorDescription } from "../";
import description from "../../../../test/test_data/description";

it("renders without defined green level", () => {
  const desc = description[0];
  const { container } = render(
    <table>
      <tbody>
        <tr>
          <IndicatorDescription description={desc} />
        </tr>
      </tbody>
    </table>,
  );
  expect(container).toMatchSnapshot();
});

it("renders with defined green level", () => {
  const desc = description[1];
  const { container } = render(
    <table>
      <tbody>
        <tr>
          <IndicatorDescription description={desc} />
        </tr>
      </tbody>
    </table>,
  );
  expect(container).toMatchSnapshot();
});

it("renders with non-complete year", () => {
  const desc = description[1];
  const { container } = render(
    <table>
      <tbody>
        <tr>
          <IndicatorDescription description={desc} lastCompleteYear={1979} />
        </tr>
      </tbody>
    </table>,
  );
  expect(container).toMatchSnapshot();
});
