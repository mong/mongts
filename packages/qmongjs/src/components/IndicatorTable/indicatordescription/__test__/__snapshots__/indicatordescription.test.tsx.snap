// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with defined green level 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="quality_indicator"
        >
          <div
            class="quality_indicator_name"
          >
            <h1>
              HbA1c &lt; 7,0 %
            </h1>
          </div>
          <div
            class="qi_long_description"
          >
            <p>
              Andel pasienter med HbA1c &lt; 7,0 %
            </p>
          </div>
          <div
            class="desired_target_level"
          >
            <h4>
              ØNSKET MÅLNIVÅ: 
              ≥
               
              40 %
            </h4>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders with non-complete year 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="quality_indicator"
        >
          <div
            class="quality_indicator_name"
          >
            <h1>
              HbA1c &lt; 7,0 %
            </h1>
          </div>
          <div
            class="qi_long_description"
          >
            <p>
              Andel pasienter med HbA1c &lt; 7,0 %
            </p>
          </div>
          <div
            class="desired_target_level"
          >
            <h4>
              ØNSKET MÅLNIVÅ: 
              ≥
               
              40 %
            </h4>
            <h4
              style="color: rgb(227, 7, 19);"
            >
              Data for valgt år er ikke komplett. Siste komplette år er
               
              1979
            </h4>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders without defined green level 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="quality_indicator"
        >
          <div
            class="quality_indicator_name"
          >
            <h1>
              HbA1c ≥ 9,0 %
            </h1>
          </div>
          <div
            class="qi_long_description"
          >
            <p>
              Andel pasienter med HbA1c &gt; 9,0
            </p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
